### DISCUSSION TITLE
Sendgrid not delivering emails when sent from a Vercel deployment.

### DISCUSSION THREAD

**Discussion by SelvinM** [vercel/community#116] [Help] [answered]
Hello, I'm using graphql in conjunction with @sendgrid/mail library on node js for email confirmation after account registration but emails aren't being delivered or are taking too long to arrive. Sometimes the email will arrive after several calls to the mutation function.

I was wondering if this could be an issue with the vercel platform since the emails seem to be working fine on the local environment. 

The issue only occurs on deployments. I'm basically doing the same as in this post https://vercel.com/guides/deploying-nextjs-nodejs-and-sendgrid-with-vercel. However, I noticed that the example deployment provided by the post doesn't seem to be working either https://nextjs-sendgrid-nodejs.now-examples.now.sh/ even though the api call is responding with status 200.

I created this repo so you can reproduce the issue if you want https://github.com/SelvinM/sendgrid-with-graphql-and-typescript-nextjs-example

Here is the repo deployed on vercel
https://sendgrid-with-graphql-and-typescript-nextjs-example.now.sh/

To reproduce the repo you'll have to create your own sendgrid API key and authenticate your own sender email on Sendgrid. Then you'll have to configure both of those things as evironment variables. 
The environment variables should be named as such:
SENDGRID_API_KEY="your sendgrid api key"
SENDER_EMAIL="configured sender email"

TL;DR: If you try sending an email locally, the email will be quickly delivered. However, if you try sending the email from a Vercel deployment, the email will be sent... but it will have trouble getting delivered.

**Comment by dominik2323** [comment, marked as answer]
I have exactly the same issue. In the local env it works like a charm, but in production, it sometimes takes too long to deliver and sometimes it fails completly.

**Reply by SelvinM** [reply]
I know man! And I still haven't figured out why it's happening. I sent a ticket to vercel support. They replied by telling me I have to provide a way to reproduce the problem. I did that 4 days ago and still haven't gotten an answer. It's a shame really. I love their platform but if I can't use sendgrid on their platform I guess I'l have to find some other hosting service.

**Reply by dominik2323** [reply]
Hey, I actually figured it out. It was super rockie mistake – I forgot to await an promise. However I don't know why it sometimes worked in production and every single time on local env. 

If this is not your case, than I found these articles:
[Sendgrid troubleshooting with delays](https://sendgrid.com/docs/ui/account-and-settings/troubleshooting-delays-and-latency/)
[Integrating SendGrid into Next.js app](https://dev.to/markdrew53/integrating-sendgrid-with-next-js-4f5m)

I hope this will help you.

**Reply by styfle** [reply]
You must await any async task or else you could end up with undefined behavior where sometimes it completes before the function returns and sometimes it does not.

See also:
- Limits: https://vercel.com/docs/v2/platform/limits#streaming-responses
- Sendgrid example: https://vercel.com/guides/deploying-nextjs-nodejs-and-sendgrid-with-vercel

**Reply by visalpii** [reply]
> I know man! And I still haven't figured out why it's happening. I sent a ticket to vercel support. They replied by telling me I have to provide a way to reproduce the problem. I did that 4 days ago and still haven't gotten an answer. It's a shame really. I love their platform but if I can't use sendgrid on their platform I guess I'l have to find some other hosting service.

I know this is quite the late answer haha but I had the same issue and the above answers lead me in the right direction - I was missing an await clause in front of my mail.send() function that SendGrid provides.

**Reply by ChristophAckerl** [reply]
> > I know man! And I still haven't figured out why it's happening. I sent a ticket to vercel support. They replied by telling me I have to provide a way to reproduce the problem. I did that 4 days ago and still haven't gotten an answer. It's a shame really. I love their platform but if I can't use sendgrid on their platform I guess I'l have to find some other hosting service.
> 
> I know this is quite the late answer haha but I had the same issue and the above answers lead me in the right direction - I was missing an await clause in front of my mail.send() function that SendGrid provides.

Same here. Just had to put await right infront of sgMail.send().

**Reply by davevilela** [reply]
Any updates? I'm still having this issue.
My code seems fine, and it works like a charm in dev mode.

**Reply by ale-strada** [reply]
> ¿Alguna actualización? Todavía tengo este problema. Mi código parece estar bien, y funciona a las mil maravillas en modo desarrollador.

a mi tambien me paso, me faltaba un await. No en la funcion sgMail.send(), sino en otra que utilizaba a esta

**Reply by binaryme** [reply]
Por acá fue lo mismo, pasé de 
`sgMail
    .send(msg)
    .then((response) => {
      console.log("response: ", response)
      console.log("email sent")
    })
    .catch((error) => {
      console.log("error: ", JSON.stringify(error))
      console.log("email not sent")
    })` a
        
    ` try {
    const sended = await sgMail.send(msg)
    console.log("sended: ", sended)
    return { success: true }
  } catch (error) {
    console.error(error)
    return { succes: false }
  }`
    
   aunque no se si lo que lo solucionó fue el cambio de sintaxis o el hecho de que ahora coloqué un return.

**Reply by gabrielmoris** [reply]
I just came to say thank you for your help.

**Reply by jlopezvaldez** [reply]
Thanks for this! Had no idea why it would work just fine locally, but not in prod.
